qutebrowser/components/braveadblock.py

```diff
@@ -0,0 +1,325 @@
+    cmdutils,
+    hook,
+    config,
+    message,
+    downloads,
+    interceptor,
+    apitypes,
+    qtutils,
+)
+from qutebrowser.api.interceptor import ResourceType
+
+if typing.TYPE_CHECKING:
+    import adblock
+
+
+logger = logging.getLogger("network")
+_ad_blocker = typing.cast(typing.Optional["BraveAdBlocker"], None)
+
+
+def _is_whitelisted_url(url: QUrl) -> bool:
+    """Check if the given URL is on the adblock whitelist."""
+    for pattern in config.val.content.blocking.whitelist:
+        if pattern.matches(url):
+            return True
+
+    # Prevent a block-list from being able to block itself, otherwise you
+    # couldn't update it.
+    blocklists = (
+        config.val.content.blocking.adblock.lists
+        + config.val.content.blocking.hosts.lists
+    )
+    if blocklists is not None:
+        if url in blocklists:
+            return True
+
+    return False
+
+
+_RESOURCE_TYPE_STRINGS = {
+    ResourceType.main_frame: "main_frame",
+    ResourceType.sub_frame: "sub_frame",
+    ResourceType.stylesheet: "stylesheet",
+    ResourceType.script: "script",
+    ResourceType.image: "image",
+    ResourceType.font_resource: "font",
+    ResourceType.sub_resource: "sub_frame",
+    ResourceType.object: "object",
+    ResourceType.media: "media",
+    ResourceType.worker: "other",
+    ResourceType.shared_worker: "other",
+    ResourceType.prefetch: "other",
+    ResourceType.favicon: "image",
+    ResourceType.xhr: "xhr",
+    ResourceType.ping: "ping",
+    ResourceType.service_worker: "other",
+    ResourceType.csp_report: "csp_report",
+    ResourceType.plugin_resource: "other",
+    ResourceType.preload_main_frame: "other",
+    ResourceType.preload_sub_frame: "other",
+    ResourceType.unknown: "other",
+    None: "",
+}
+
+
+def resource_type_to_string(resource_type: typing.Optional[ResourceType]) -> str:
+    return _RESOURCE_TYPE_STRINGS.get(resource_type, "other")
+
+
+# TODO: Move this code somewhere so that `adblock.py` can make use of it too.
+class _FakeDownload(downloads.TempDownload):
+
+    """A download stub to use on_download_finished with local files."""
+
+    def __init__(
+        self, fileobj: typing.IO[bytes]  # pylint: disable=super-init-not-called
+    ) -> None:
+        self.fileobj = fileobj
+        self.successful = True
+
+
+class BraveAdBlocker:
+
+    """Manage blocked hosts based from /etc/hosts-like files.
+
+    Attributes:
+        _in_progress: The DownloadItems which are currently downloading.
+        _done_count: How many files have been read successfully.
+        _has_basedir: Whether a custom --basedir is set.
+        _cache_path: The path of the adblock engine cache file
+        _engine_factory: A function that can be called to create an empty
+                         adblocking engine.
+        _engine: Brave ad-blocking engine.
+    """
+
+    def __init__(
+        self,
+        *,
+        engine_factory: typing.Callable[[], "adblock.Engine"],
```

Comment:
Why you need to pass an `engine_factory` from the outside, is that just an artifact from when the imports in this module worked differently?

Wouldn't it suffice to do `self._engine = adblock.Engine([])` here directly (and in `adblock_update`)? You'd need to change the imports to something like:

```python
try:
    import adblock
except ImportError:
    adblock = None  # type: ignore[assignment]
```

and can then just do `if adblock is not None: ...` in `init`.